src/sysupdate/sysupdated.c

```diff
@@ -1645,65 +1646,39 @@ static Manager *manager_free(Manager *m) {
 DEFINE_TRIVIAL_CLEANUP_FUNC(Manager *, manager_free);
 
 static int manager_on_notify(sd_event_source *s, int fd, uint32_t revents, void *userdata) {
-        char buf[NOTIFY_BUFFER_MAX+1];
-        struct iovec iovec = {
-                .iov_base = buf,
-                .iov_len = sizeof(buf)-1,
-        };
-        CMSG_BUFFER_TYPE(CMSG_SPACE(sizeof(struct ucred))) control;
-        struct msghdr msghdr = {
-                .msg_iov = &iovec,
-                .msg_iovlen = 1,
-                .msg_control = &control,
-                .msg_controllen = sizeof(control),
-        };
-        struct ucred *ucred;
         Manager *m = ASSERT_PTR(userdata);
-        Job *j;
-        ssize_t n;
-        char *version, *progress, *errno_str, *ready;
-
-        n = recvmsg_safe(fd, &msghdr, MSG_DONTWAIT|MSG_CMSG_CLOEXEC);
-        if (ERRNO_IS_NEG_TRANSIENT(n))
-                return 0;
-        if (n == -ECHRNG) {
-                log_warning_errno(n, "Got message with truncated control data (unexpected fds sent?), ignoring.");
-                return 0;
-        }
-        if (n == -EXFULL) {
-                log_warning_errno(n, "Got message with truncated payload data, ignoring.");
-                return 0;
-        }
-        if (n < 0)
-                return (int) n;
+        int r;
 
-        cmsg_close_all(&msghdr);
+        assert(fd >= 0);
 
-        ucred = CMSG_FIND_DATA(&msghdr, SOL_SOCKET, SCM_CREDENTIALS, struct ucred);
-        if (!ucred || ucred->pid <= 0) {
-                log_warning("Got notification datagram lacking credential information, ignoring.");
+        _cleanup_(pidref_done) PidRef sender_pidref = PIDREF_NULL;
+        _cleanup_free_ char *buf = NULL;
+        r = notify_recv(fd, &buf, /* ret_ucred= */ NULL, &sender_pidref);
+        if (r == -EAGAIN)
                 return 0;
-        }
+        if (r < 0)
+                return log_warning_errno(r, "Failed to receive notification message: %m");
 
+        Job *j;
         HASHMAP_FOREACH(j, m->jobs) {
-                pid_t pid;
-                assert_se(sd_event_source_get_child_pid(j->child, &pid) >= 0);
+                PidRef child_pidref = PIDREF_NULL;
+
+                r = event_source_get_child_pidref(j->child, &child_pidref);
+                if (r < 0)
+                        return log_warning_errno(r, "Failed to get child pidref: %m");
```

Comment:
Then, this can remain assert_se

Reply:
hmm, i am not too much of a fan of encoding too much knowledge about the implementation of event_source_get_child_pidref() here and all the functions it might call that might fail. hence, I think we should continue to catch the erro properly here.